Pass files to golint one at a time

cblecker opened the pull request:
**What this PR does / why we need it**:
When we pass multiple files into golint at once, and golint detects a fatal error in one of them, it will actually exit and not process any more of the files passed. For large packages, this increases the chance that golint will exit.

This change will, instead of golinting once per package, will use xargs to run each file through golint individually.

**Special notes for your reviewer**:
Out of interest, I timed the difference between using find and a regex to exclude the generated files, but `ls | grep` was about 15-20 seconds faster throughout the entire duration of the run (about 2 minutes).

**Release note**:
```release-note
NONE
```

cblecker commented:
/cc @nikhita 

Still need to write stuff up on this one, but figured I might as well throw it in front of CI while I grab some lunch.

nikhita commented on hack/verify-golint.sh:
```diff
@@ -74,7 +74,7 @@ for p in "${all_packages[@]}"; do
 	# Packages with a corresponding foo_test package will make golint fail
 	# with a useless error. Just ignore that, see golang/lint#68.
-	failedLint=$(golint "$p"/*.go 2>/dev/null)
+	failedLint=$(ls "$p"/*.go | egrep -v "(zz_generated.*.go|generated.pb.go|generated.proto|types_swagger_doc_generated.go)" | xargs -L1 golint 2>/dev/null)
```
@cblecker thank you!

This line looks good. If CI likes it overall, lgtm. :+1:

nikhita commented:
@cblecker just a heads up that once https://github.com/kubernetes/kubernetes/pull/67635 merges, the list would need to be updated again.

nikhita commented:
/cc @spxtr 
since you had updated this file previously

cblecker committed:
Convert tabs to spaces

cblecker commented:
/retest

cblecker commented:
This is what the failures look like now that it's actually looking at all files:
https://gist.github.com/cblecker/55b0f86d93b5b7c6d6857d68f35668fc

cblecker committed:
Add packages to golint_failures

nikhita commented on hack/verify-golint.sh:
```diff
@@ -46,88 +46,90 @@ array_contains () {
+  # and not just the ones for the current platform. This also will ensure that
+  # _test files are linted.
+  # Generated files are ignored, and each file is passed through golint
+  # individually, as if one file in the package contains a fatal error, golint
```
Can we add some more details on why a package might contain a fatal error? Something like: `as if one file in the package contains a fatal error due to the presence of a corresponding foo_test package`.

The `foo_test` line in the script helped me find the problem in a second. :)

nikhita commented on hack/verify-golint.sh:
```diff
@@ -46,88 +46,90 @@ array_contains () {
-	fi
+  # Run golint on package/*.go file explicitly to validate all go files
+  # and not just the ones for the current platform. This also will ensure that
+  # _test files are linted.
```
nit: _test.go files?

nikhita commented:
Adding some more details so that if someone comes across this, it's easier for them to understand the line of thought. :)

> When we pass multiple files into golint at once, and golint detects a fatal error in one of them, it will actually exit and not process any more of the files passed.

This is due to an existing golint bug: https://github.com/golang/lint/issues/68#issuecomment-55737653. If we lint a package using `golint p/*.go` and the package `p` contains a test files with the package name `p_test`, this gives a fatal error, something like:

```
staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured_test.go is in package unstructured_test, not unstructured
```
and skips the linting. Current behaviour was to ignore this error, which also meant that we ignored the linting of these packages all together.

Also, it's not possible to lint the whole package together through `golint p` because that would not lint `_test.go` files due to another golint bug: https://github.com/golang/lint/issues/200.

> This change will, instead of golinting once per package, will use xargs to run each file through golint individually.

And so we use xargs to make sure we are linting all files and don't hit any errors. :sparkles:

cblecker committed:
Fix golint command to only pass a single *.go file at a time

cblecker commented:
@nikhita Addressed your feedback. PTAL!

Also thanks for the amazing write up ✨ 🗒 ! I linked to this PR in the comments to so future bash spelunkers will be able to find it! :bash:

nikhita commented:
> I linked to this PR in the comments to so future bash spelunkers will be able to find it! 

:tada: 

/lgtm
/retest
